File: server/src/intTest/java/org/tctalent/server/repository/db/HelpLinkSettingsSpecificationIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getHelpLink;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedCountry;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.tctalent.server.model.db.HelpLink;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;
import org.tctalent.server.request.helplink.SearchHelpLinkRequest;

public class HelpLinkSettingsSpecificationIntTest extends BaseDBIntegrationTest {

  @Autowired
  private HelpLinkRepository repo;
  @Autowired
  private CountryRepository countryRepo;
  private HelpLink helpLink;
  private Specification<HelpLink> spec;
  private SearchHelpLinkRequest request;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    helpLink = getHelpLink();
    helpLink.setCountry(getSavedCountry(countryRepo));
    repo.save(helpLink);
    assertTrue(helpLink.getId() > 0);
    request = new SearchHelpLinkRequest();
  }

  @Test
  public void testKeyword() {
    request.setKeyword(helpLink.getLabel());
    spec = HelpLinkSettingsSpecification.buildSearchQuery(request);
    List<HelpLink> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(helpLink.getId(), results.getFirst().getId());
  }

  @Test
  public void testKeywordWithLink() {
    request.setKeyword(helpLink.getLink());
    spec = HelpLinkSettingsSpecification.buildSearchQuery(request);
    List<HelpLink> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(helpLink.getId(), results.getFirst().getId());
  }

  @Test
  public void testKeywordFail() {
    request.setKeyword("NOTHING");
    spec = HelpLinkSettingsSpecification.buildSearchQuery(request);
    List<HelpLink> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }

  @Test
  public void testWithCountry() {
    assert helpLink.getCountry() != null;
    request.setCountryId(helpLink.getCountry().getId());
    spec = HelpLinkSettingsSpecification.buildSearchQuery(request);
    List<HelpLink> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    List<Long> ids = results.stream().map(HelpLink::getId).toList();
    assertTrue(ids.contains(helpLink.getId()));
  }

  @Test
  public void testNoCountry() {
    request.setCountryId(0L);
    spec = HelpLinkSettingsSpecification.buildSearchQuery(request);
    List<HelpLink> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/HelpLinkSettingsSpecificationIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.HelpLink
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getHelpLink
import org.tctalent.server.repository.db.integrationhelp.getSavedCountry
import org.tctalent.server.request.helplink.SearchHelpLinkRequest

class HelpLinkSettingsSpecificationIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: HelpLinkRepository
  @Autowired lateinit var countryRepo: CountryRepository
  lateinit var helpLink: HelpLink

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    helpLink = getHelpLink().apply { country = getSavedCountry(countryRepo) }
    repo.save(helpLink)
    assertTrue { helpLink.id > 0 }
  }

  @Test
  fun `test keyword`() {
    val request = SearchHelpLinkRequest().apply { keyword = helpLink.label }
    val spec = HelpLinkSettingsSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    assertEquals(1, result.size)
    assertEquals(helpLink.id, result.first().id)
  }

  @Test
  fun `test keyword with link`() {
    val request = SearchHelpLinkRequest().apply { keyword = helpLink.link }
    val spec = HelpLinkSettingsSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    assertEquals(1, result.size)
    assertEquals(helpLink.id, result.first().id)
  }

  @Test
  fun `test keyword fail`() {
    val request = SearchHelpLinkRequest().apply { keyword = "NOTHING" }
    val spec = HelpLinkSettingsSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue { result.isEmpty() }
  }

  @Test
  fun `test with country`() {
    val request = SearchHelpLinkRequest().apply { countryId = helpLink.country?.id ?: fail() }
    val spec = HelpLinkSettingsSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    val ids = result.map { it.id }
    assertTrue { ids.contains(helpLink.id) }
  }

  @Test
  fun `test no country`() {
    val request = SearchHelpLinkRequest().apply { countryId = 0 }
    val spec = HelpLinkSettingsSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isEmpty())
  }
}
